internal/lifecycle/readme.go

```diff
@@ -0,0 +1,127 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lifecycle
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/util"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+// are to be used by this program for building and deploying the sample.
+var codeTag = "sst-run-linuxmacos"
+
+// parseREADME parses a README file with the given name. It reads terminal commands surrounded by one of the codeTags
+// listed above and loads them into a Lifecycle. In the process, it replaces the Cloud Run service name and Container
+// Registry tag with the provided inputs.
+func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+
+	var lifecycle Lifecycle
+
+	inCode := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if strings.Contains(line, "```") && inCode {
+			inCode = false
+		} else if strings.Contains(line, codeTag) {
+			scanner.Scan()
+			startCodeBlockLine := scanner.Text()
+
+			if !strings.Contains(startCodeBlockLine, "```") {
+				return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: incorrect format")
+			}
+
+			inCode = true
+		} else if inCode {
+			for line[len(line)-1] == '\\' {
+				line = line[:len(line)-1]
+
+				scanner.Scan()
+				line = line + strings.TrimSpace(scanner.Text())
+			}
+
+			line = os.ExpandEnv(line)
+			line = replaceGCRURL(line, gcrURL)
+			line = replaceServiceName(line, serviceName)
+			sp := strings.Split(line, " ")
+
+			var cmd *exec.Cmd
+			if strings.Contains(line, "gcloud") {
+				a := append(util.GcloudCommonFlags, sp[1:]...)
+				cmd = exec.Command("gcloud", a...)
+			} else {
+				cmd = exec.Command(sp[0], sp[1:]...)
+			}
+
+			lifecycle = append(lifecycle, cmd)
+		}
+	}
+
+	if inCode {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: incorrect format")
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+	}
+
+	if len(lifecycle) == 0 {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: no commands found")
+	}
+
+	return lifecycle, nil
+}
+
+// replaceServiceName takes a terminal command string as input and replaces the Cloud Run service name, if any.
+// It detects whether the command is a gcloud run command and replaces the last argument that isn't a flag
+// with the input service name.
+func replaceServiceName(command, serviceName string) string {
+	containsGcloud, _ := regexp.MatchString(`\bgcloud\b`, command)
```

Comment:
Probably worth a comment on why we ignore these errors

Reply:
Will add a comment. Also an explanation here:

The only scenario in which the `regexp.MatchString` function will return an error is if the regex fails to compile. But in our case, since we're using the same two pieces of regex each time and we know that they're valid, it seemed like we could safely ignore these errors.

`regexp.MatchString` source: https://github.com/golang/go/blob/master/src/regexp/regexp.go#L548 .

Given this, I don't know if it's acceptable to be ignoring the errors or if we should be catching the errors for good measure anyways.

Reply:
A better alternative here would be to save the regex as a constant, so it's calculated at compile time. You can use `regex.MustCompile()` and then `(*Regexp).MatchString`. This will help you avoid the case where someone modifies the regex and you don't find out it's invalid until your compiler runs.

Reply:
Sorry, do you mean something like this?

```
r1 := regexp.MustCompile(`\bgcloud\b`)
r2 := regexp.MustCompile(`\brun\b`)

containsGcloud:= r1.MatchString(command)
containsRun := r2.MatchString(command)
```

Reply:
Yup!

Reply:
Resolved, thank you.